extensions/ql-vscode/src/helpers.ts

```diff
@@ -50,10 +78,58 @@ export function withProgress<R>(
         const increment = 100 * (step - progressAchieved) / maxStep;
         progressAchieved = step;
         progress.report({ message, increment });
-      }, token);
+      }, token, ...args);
     });
 }
 
+/**
+ * A generic wrapper for commands. This wrapper adds error handling and progress monitoring
+ * for any command.
+ *
+ * There are two ways to invoke the command task: with or without a progress monitor
+ * If progressOptions are passed in, then the command task will run with a progress monitor
+ * Otherwise, no progress monitor will be used.
+ *
+ * If a task is run with a progress monitor, the first two arguments to the task are always
+ * the progress callback, and the cancellation token. And this is followed by any extra command arguments
+ * (eg- selection, multiselection, ...).
+ *
+ * If there is no progress monitor, then only extra command arguments are passed in.
+ *
+ * @param commandId The ID of the command to register.
+ * @param task The task to run. If passing taskOptions, then this task must be a `ProgressTask`.
+ * @param progressOptions Optional argument. If present, then the task is run with a progress monitor
+ *                    and cancellation token, otherwise it is run with no arguments.
+ */
+export function commandRunner<R>(
+  commandId: string,
+  task: ProgressTask<R> | NoProgressTask,
+  progressOptions?: ProgressOptions
+): Disposable {
+  return commands.registerCommand(commandId, async (...args: any[]) => {
+    try {
+      if (progressOptions) {
+        await withProgress(progressOptions, task as ProgressTask<R>, ...args);
+      } else {
+        await (task as ((...args: any[]) => Promise<any>))(...args);
+      }
+    } catch (e) {
+      if (e instanceof UserCancellationException) {
+        // User has cancelled this action manually
+        if (e.silent) {
+          logger.log(e.message);
+        } else {
+          showAndLogWarningMessage(e.message);
+        }
+      } else if (e instanceof Error) {
+        showAndLogErrorMessage(e.message);
+      } else {
+        throw e;
```

Comment:
Should we always handle this (show and log) to avoid uncaught promise rejections at the top level?

Reply:
Our code should _always_ be throwing errors. If a non-error is thrown, then that is just...strange and something unexpected is happening. I don't feel strongly about this since it _shouldn't_ happen. I can remove.

Reply:
Need to handle the case where `e.message` is undefined.